Sprint 8 solution time and duration

fcevt opened the pull request:
Выполнил задание 8го спринта

fcevt committed:
Выполнить задание 8го спринта

fcevt committed:
Выполнить задание 8го спринта

pattyku commented on src/com/yandex/tracker/servise/InMemoryTaskManager.java:
```diff
@@ -234,4 +247,138 @@ public void updateEpicStatus(int epicId) {
+        return prioritizedTasks;
+    }
+
+    //удаление задачи из отсортированного списка при удалении задачи по id
```
Не нужно четыре отдельных метода, достаточно одного и вызываешь его везде где требуется
```
private void delete(Task task) {
	prioritizedTasks.remove(task);
}
```

pattyku commented on src/com/yandex/tracker/servise/InMemoryTaskManager.java:
```diff
@@ -234,4 +247,138 @@ public void updateEpicStatus(int epicId) {
+            if (getPrioritizedTasks().stream()
+                    .anyMatch(t -> timeIntersectionSearch(task, t))) {
+                if (task instanceof Subtask) {
+                    deleteSubtaskById(task.getId());
```
Не вызываем публичные методы в приватных методах. В этом методе нужно добавить в prioritizedTasks, но не удаляем. Старайся придерживаться правила: метод содержит только ту логику, которая содержится в его названии. 
А в методах deleteSubtaskById и deleteTaskById делается много дополнительной логики, которая здесь не нужна

Пример метода для изучения

```
private void add(Task newTask) {
		prioritizedTasks.stream()
				.filter(existingTask -> isOverlapping(newTask, existingTask))
				.findFirst()
				.ifPresentOrElse(
						// Если нашлась задача с пересекающимся временем, то генерируем исключение
						overlappedTask -> {
							String message = "Задача пересекаются с id=" + overlappedTask.getId() +
									" c " + overlappedTask.getStartTime() + " по " +
									overlappedTask.getEndTime();
							throw new TaskValidationException(message);
						},
						// если задач пересекающихся по времени нет, то добавляем новую задачу
						() -> prioritizedTasks.add(newTask)
				);
	}

```

pattyku commented on src/com/yandex/tracker/servise/InMemoryTaskManager.java:
```diff
@@ -234,4 +247,138 @@ public void updateEpicStatus(int epicId) {
+    }
+
+    //получение сортированного по времени списка задач
+    public List<Task> getPrioritizedTasks() {
```
Также этот метод нужно вынести в интерфейс `TaskManager`

pattyku commented on src/com/yandex/tracker/servise/InMemoryTaskManager.java:
```diff
@@ -234,4 +247,138 @@ public void updateEpicStatus(int epicId) {
+    }
+
+    //обновление задач в мапах и трисете для методов update
+    private void updatePrioritizedTasks(Task task) {
```
Очень сложный метод. Его сложно читать и будет очень сложно поддерживать. Не всегда объединять методы это хорошо.
Предлагаю удалить этот метод и нужную логику перенести в методы обновления задач
Пример для Task

```
public void updateTask(Task task) {
		final int id = task.getId();
		final Task savedTask = tasks.get(id);
		if (savedTask == null) {
			return;
		}
		deletePrioritizedTasks(savedTask);
		tasks.put(id, task);
		addPrioritizedTasks(task);
} 
```

Аналогично для подзадач, только там нужно сделать дополнительную логику для эпика

fcevt committed:
Заменить 4 метода удаления из трисета на 1 и изменить логику где они использовались.
Вынести метод получения сортированного списка интерфейс TaskManager.
Изменить методы создания задач в т.ч заменить в них  возвращаемое значение на optional чтобы не возвращать null, если задача не создана.
Изменить метод добавления в трисет.
Изменить метод обновления задач на два в зависимости от типа.
Изменить под все это тесты.

pattyku commented on src/com/yandex/tracker/servise/InMemoryTaskManager.java:
```diff
@@ -1,25 +1,34 @@
+        prioritizedTasks = new TreeSet<>();
     }
 
+    public HistoryManager getBrowsingHistory() {
```
По ТЗ такой метод не нужен и нигде его не используешь, удаляем

pattyku commented on src/com/yandex/tracker/servise/InMemoryTaskManager.java:
```diff
@@ -93,12 +101,14 @@ public void removeEpics() {
         for (Epic epic : epics.values()) {
             epic.removeListOfSubtasks();
             epic.setStatus(TaskStatus.NEW);
             updateEpic(epic);
```
Здесь нужно было не эпик обновлять, а его статус. Вызываем метод updateEpicStatus()

pattyku commented on src/com/yandex/tracker/servise/InMemoryTaskManager.java:
```diff
@@ -123,37 +133,80 @@ public Subtask getSubtasksById(int id) {
+        try {
+            addToPrioritizedTasks(task);
+        } catch (TimeIntersectionException e) {
+            System.out.println(e.getMessage());
```
Оборачивать в try-catch не нужно. Пользовательский ввод/вывод используем в классе Main. В менеджерах не нужно.  System.out.println в этом классе везде нужно удалить

pattyku commented on src/com/yandex/tracker/servise/InMemoryTaskManager.java:
```diff
@@ -123,37 +133,80 @@ public Subtask getSubtasksById(int id) {
         subtasks.put(id, subtask);
         Epic epic = epics.get(epicId);
         epic.addSubtaskToList(subtask.getId());
         updateEpic(epic);
```
updateEpic(epic); - удаляем

pattyku commented on src/com/yandex/tracker/servise/InMemoryTaskManager.java:
```diff
@@ -123,37 +133,80 @@ public Subtask getSubtasksById(int id) {
         id++;
         subtask.setId(id);
         subtask.setEpicId(epicId);
+        try {
```
актуален комментарий выше

pattyku commented on src/com/yandex/tracker/servise/InMemoryTaskManager.java:
```diff
@@ -123,37 +133,80 @@ public Subtask getSubtasksById(int id) {
+            if (savedTask == null) {
+                throw new NonExistingTaskException("Задача не обновлена т.к задачи с id=" + id + " не существует");
+            }
+            if (task.getStartTime().isEqual(savedTask.getStartTime()) && //если в метод передали задачу с
```
У тебя есть методы для добавления и удаления из prioritizedTasks() и их нужно использовать здесь


```
@Override
    public void updateTask(Task task) {
        final int id = task.getId();
        final Task savedTask = tasks.get(id);

        if (savedTask == null) {
                throw new NonExistingTaskException("Задача не обновлена т.к задачи с id=" + id + " не существует");
        }
            
        deleteTaskFromPrioritizedTasks(savedTask);
	tasks.put(task.getId(), task);
	addToPrioritizedTasks(task);
    }
```

Этот же комментарий актуален и для метода updateSubtask

pattyku commented on src/com/yandex/tracker/servise/InMemoryTaskManager.java:
```diff
@@ -234,4 +318,118 @@ public void updateEpicStatus(int epicId) {
+    }
+
+    //обновление задач в мапах и трисете для методов update
+    private void updatePrioritizedTasks(Task task) {
```
Этот метод уже не используется нужно удалить

pattyku commented on src/com/yandex/tracker/servise/InMemoryTaskManager.java:
```diff
@@ -234,4 +318,118 @@ public void updateEpicStatus(int epicId) {
+
+    //удаление подзадачи по id из отсортированного списка при удалении эпика по id
+    private void deleteTaskFromPrioritizedTasks(int id) {
+        prioritizedTasks.removeIf(task -> task.getId() == id);
```
`removeIf `используется когда нужно удалить несколько элементов из коллекции, которые удовлетворяет определенному условию. Например все задачи у которых статус равен done. 

здесь удаление по id и не может быть двух задач одинаковым id в Set, потому что не допускается дублирования значений. Поэтому нужно использовать метод `remove()` который удаляет один объект из набора

pattyku commented on src/com/yandex/tracker/servise/InMemoryTaskManager.java:
```diff
@@ -206,7 +290,7 @@ public void updateEpicStatus(int epicId, TaskStatus newStatus) {
 
     @Override
     public void updateEpicStatus(int epicId) {
```
Тут пропустила `updateEpicStatus() ` используется только внутри этого класса и не должен быть доступен во вне, поэтому private вместо public

fcevt committed:
Удалить не нужные getBrowsingHistory и updatePrioritizedTasks.
Заменить обновление эпика на обновление статуса.
Убрать везде оборачивание в try-catch.
Изменить методы update на твой вариант.
Заменить removeIf на remove.
Удалить updateEpicStatus из интерфейса, чтобы сделать их private.
Из-за выше перечисленного частично изменить логику работы методов create и addToPrioritizedTasks
и часть тестов.

fcevt commented:
Привет, все сделал, но не понял чем были плохи мои методы update, да они были больше но зато с ними можно было обновлять задачу не только по статусу но и по времени с проверкой пересечения. И еще не понял почему нужно убрать try-catch. у меня на catch было завязано немного логики и в их отсутствии приходится в методах create делать дополнительные проверки. Объясни если не сложно тут или в пачке krikunovalexandr1990

pattyku commented on src/com/yandex/tracker/servise/InMemoryTaskManager.java:
```diff
@@ -123,38 +128,57 @@ public Subtask getSubtasksById(int id) {
+        addToPrioritizedTasks(task);
+        if (!prioritizedTasks.contains(task)) {
+            return Optional.empty();
+        }
```
По ТЗ добавление задачи в prioritizedTasks  не должно влиять на добавление в tasks.put. Поэтому и  попросила убрать блок  try-catch. Здесь делаешь тоже самое но уже с условием. Не нужно возвращать return Optional.empty();

Блок из ТЗ:
"Дата начала задачи по каким-то причинам может быть не задана. Тогда при добавлении её не следует учитывать в списке задач и подзадач, отсортированных по времени начала. Такая задача не влияет на приоритет других, а при попадании в список может сломать логику работы компаратора."

pattyku commented on src/com/yandex/tracker/servise/InMemoryTaskManager.java:
```diff
@@ -123,38 +128,57 @@ public Subtask getSubtasksById(int id) {
     public void updateTask(Task task) {
+        final int id = task.getId();
+        final Task savedTask = tasks.get(id);
+        if (savedTask == null) {
```
Насчет методов update. Одна логика на все методы, не зависит от того, создается или обновляется задача. В таких кейсах код выносится в отдельный метод и используется везде. Для того чтоб не дублировать код, а также для ситуаций если в будущем что-то поменяется. Допустим через несколько месяцев будет принято решение поменять алгоритм проверки на пересечения. Нужно будет зайти в один метод и поменять логику и заработает для всех. А если будет копипаст в несколько методов, есть вероятность забыть где-то поменять и словить ошибку в дальнейшем. 

По блокам try-catch тоже самое что и в методах криейт. Независимо от того добавилась задача в prioritizedTasks, обновлять ее в tasks нужно.

Проверку ниже можно было оставить. Если время не изменилось, не нужно проходится по сету в поисках пересечений. Экономия памяти при больших данных. Тут отлично

```
            if (task.getStartTime().isEqual(savedTask.getStartTime()) && //если в метод передали задачу с
                    task.getEndTime().isEqual(savedTask.getEndTime())) { //временем начала и конца уже существующей
                prioritizedTasks.removeIf(t -> t.equals(task));    //задачи
                prioritizedTasks.add(task);
                tasks.put(task.getId(), task);
                return;
            }
```

pattyku commented:
> Привет, все сделал, но не понял чем были плохи мои методы update, да они были больше но зато с ними можно было обновлять задачу не только по статусу но и по времени с проверкой пересечения. И еще не понял почему нужно убрать try-catch. у меня на catch было завязано немного логики и в их отсутствии приходится в методах create делать дополнительные проверки. Объясни если не сложно тут или в пачке krikunovalexandr1990

Постаралась подробно объяснить комментариями, если остались вопросы пиши. Работу принимаю, постараюсь не пропустить сообщения если будут в гите или напиши в пачку

fcevt committed:
убрать возвращение optional.empty из методов create и при любом раскладе добавлять задачи и подзадачи в мапы.
Вернуть в методы update проверку на неизменность времени.

fcevt committed:
убрать возвращение optional.empty из методов create и при любом раскладе добавлять задачи и подзадачи в мапы.
Вернуть в методы update проверку на неизменность времени.

fcevt commented:
Спасибо теперь понял,  просто я был почему то убежден что при пересечении по времени и добавлять в мапу не нужно.

fcevt commented:
все указанное исправил и вернул
